### PR TITLE
Deprecate `clean_up_tokenization_spaces` for BLOOM

### PR DESCRIPTION
**Pull request by thomasw21**
# What does this PR do?

Currently in `transformers`:
```python
>>> tok.decode(tok.encode("Hello , there"))
'Hello, there' # notice the missing space between "Hello" and ","
>>> tok.decode(tok.encode("Hello , there"), clean_up_tokenization_spaces=False)
'Hello , there'
```

In order too prevent issues such as this: https://huggingface.co/bigscience/bloom/discussions/153#6397907b71eb2455d898e0a4 we suggest to add a warning, suggesting to users to use `clean_up_tokenization_spaces=False` instead.

As BLOOM tokenizer was developped in order to be lossless encoding mechanism, it should make sense to always remove that option IMO, therefore I'm suggesting to deprecate that option from BLOOM tokenizer. Other option would be to change the default to `True`.

### CODE REVIEW

File: src/transformers/models/bloom/tokenization_bloom_fast.py [hunk 1]
```diff
@@ -155,6 +165,81 @@ def _encode_plus(self, *args, **kwargs) -> BatchEncoding:
...
+
+        Returns:
+            `List[str]`: The list of decoded sentences.
+        """
+        if clean_up_tokenization_spaces:
+            logger.warning(
+                "Bloom tokenizer was built in order to have lossless encoding. Most likely, you should use"
+                " `clean_up_tokenization_spaces=False`. This flag is deprecated and will be remove in v5. The"
+                " behaviour will correspond to `clean_up_tokenization_spaces=False`"
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/bloom/tokenization_bloom_fast.py:196]:
```suggestion
            warnings.warn(
                "Bloom tokenizer was built in order to have lossless encoding. Most likely, you should use"
                " `clean_up_tokenization_spaces=False`. This flag is deprecated and will be remove in v5. The"
                " behaviour will correspond to `clean_up_tokenization_spaces=False`"
                FutureWarning,
            )
```
Let's use a `FuturWarning` here 😉

File: src/transformers/models/bloom/tokenization_bloom_fast.py [hunk 2]
```diff
@@ -155,6 +165,81 @@ def _encode_plus(self, *args, **kwargs) -> BatchEncoding:
...
+                Will be passed to the underlying model specific decode method.
+
+        Returns:
+            `str`: The decoded sentence.
+        """
+        if clean_up_tokenization_spaces:
+            logger.warning(
+                "Bloom tokenizer was built in order to have lossless encoding. Most likely, you should use"
+                " `clean_up_tokenization_spaces=False`. This flag is deprecated and will be remove in v5. The"
+                " behaviour will correspond to `clean_up_tokenization_spaces=False`"
```
Comment (ArthurZucker) [inline src/transformers/models/bloom/tokenization_bloom_fast.py:234]: Same here

**Review by ArthurZucker — APPROVED** [review]
Thanks for this!

File: src/transformers/models/bloom/tokenization_bloom_fast.py (hunk 1 above)
Comment (thomasw21) [reply src/transformers/models/bloom/tokenization_bloom_fast.py:196]: Any reason to use `warnings.warn` instead of `logger.warning`?

Comment (ArthurZucker) [reply src/transformers/models/bloom/tokenization_bloom_fast.py:196]: FuturWarning will be easier to catch once we do a cleanup and are bound to disappear, while other warnings might not

Comment (thomasw21) [reply src/transformers/models/bloom/tokenization_bloom_fast.py:196]: What I meant is can't you use `logger.warning` and have a FutureWarning? Well I just searched in the repo and all seem to be using a `warning.warn` so let's do it your way!

**Comment by LysandreJik** [comment]
Thanks for your PR! After thinking a little more about it and in terms of user experience, I'm happy to have the warning if you think the use-case is frequent and the default behavior is misleading. 

However, I'm not too sure about deprecating/updating the value in v5. I think the current behavior isn't necessarily a bug, as the argument to toggle is clearly displayed in the docs (and I have no problem with making it more prominent, such as with the warning). Switching to `False` means that we'll start diverging between BLOOM and other tokenizers (like GPT-2) which work very similarly as of now.

I'd be in favor of adding the warning mentioning to toggle it in this PR, and to wait until @sgugger is back so that we have a second opinion on the matter before mentioning that we will move it to `False` by default. Would that be ok for you @thomasw21?

**Comment by LysandreJik** [comment]
Interesting proposal, WDYT @Narsil?

**Comment by Narsil** [comment]
I think it's ok to move slowly, but touching `cleanup_tokenization_spaces` and its default are BIG changes.

Personally, I think borderline too big to migrate in V5 (it's just a really big change, that's unfortunately probably not worth the effort). 

That being said, making it modifiable on a tokenizer per tokenizer basis (so updating Bloom alone) is still Ok, and is definitely a good way forward.

Personally I would focus on this user's need first, which would be solved by implementing `return_full_text=False`, it seems the lowest hanging fruit to solve the user's need. We can move forward on the "decoder"  (or any other type of config change) later.

File: src/transformers/models/bloom/tokenization_bloom_fast.py [hunk 3]
```diff
@@ -155,6 +166,83 @@ def _encode_plus(self, *args, **kwargs) -> BatchEncoding:
...
+            skip_special_tokens (`bool`, *optional*, defaults to `False`):
+                Whether or not to remove special tokens in the decoding.
+            clean_up_tokenization_spaces (`bool`, *optional*, defaults to `True`):
+                Whether or not to clean up the tokenization spaces.
+            kwargs (additional keyword arguments, *optional*):
+                Will be passed to the underlying model specific decode method.
+
+        Returns:
+            `List[str]`: The list of decoded sentences.
+        """
```
Comment (sgugger) [inline src/transformers/models/bloom/tokenization_bloom_fast.py:191]: Remove the docstring, so it uses the docstring of the superclass (and we never have to worry about it getting outdated).

File: src/transformers/models/bloom/tokenization_bloom_fast.py [hunk 4]
```diff
@@ -155,6 +166,83 @@ def _encode_plus(self, *args, **kwargs) -> BatchEncoding:
...
+            skip_special_tokens (`bool`, *optional*, defaults to `False`):
+                Whether or not to remove special tokens in the decoding.
+            clean_up_tokenization_spaces (`bool`, *optional*, defaults to `True`):
+                Whether or not to clean up the tokenization spaces.
+            kwargs (additional keyword arguments, *optional*):
+                Will be passed to the underlying model specific decode method.
+
+        Returns:
+            `str`: The decoded sentence.
+        """
```
Comment (sgugger) [inline src/transformers/models/bloom/tokenization_bloom_fast.py:231]: Same here.

**Review by sgugger — COMMENTED** [review]
Thanks for the PR. I agree with @LysandreJik on the fact that we would probably not change the default even in v5, so the deprecation warning makes little sense. If it's possible to have it built in the fast tokenizer directly and then update repos on a case-by-case basis, I'd be more in favor of that approach.

Left comments on the docstrings in case this PR moves forward :-)

**Comment by thomasw21** [comment]
Okay so in terms of actions:
 - [ ] Assume it's not a `transformers` bug but a `text-generation-inference` bug right now.
 - [ ] Start thinking of a way to support `clean_up_tokenization_spaces` and `skip_special_tokens` in the tokenizer directly? Typically you want in order of priority: `user defined argument`, `tokenizer specific config`, `methods default`

**Pull request closed without merging**